src/librustc_mir/transform/inline.rs

```diff
@@ -94,33 +95,39 @@ impl Inliner<'tcx> {
 
         loop {
             local_change = false;
-            while let Some(callsite) = callsites.pop_front() {
+            'callsites: while let Some(callsite) = callsites.pop_front() {
                 debug!("checking whether to inline callsite {:?}", callsite);
                 if !self.tcx.is_mir_available(callsite.callee) {
                     debug!("checking whether to inline callsite {:?} - MIR unavailable", callsite);
                     continue;
                 }
 
-                let self_node_id = self.tcx.hir().as_local_node_id(self.source.def_id()).unwrap();
-                let callee_node_id = self.tcx.hir().as_local_node_id(callsite.callee);
-
-                let callee_body = if let Some(callee_node_id) = callee_node_id {
+                // This may result in a cycle since the callee MIR is from this crate
+                // and we be optimizing it right now in this very same query stack.
+                if callsite.callee.is_local() {
                     // Avoid a cycle here by only using `optimized_mir` only if we have
-                    // a lower node id than the callee. This ensures that the callee will
-                    // not inline us. This trick only works without incremental compilation.
-                    // So don't do it if that is enabled.
-                    if !self.tcx.dep_graph.is_fully_enabled()
-                        && self_node_id.as_u32() < callee_node_id.as_u32()
-                    {
-                        self.tcx.optimized_mir(callsite.callee)
-                    } else {
-                        continue;
+                    // no mutual recursion. This ensures that the callee will
+                    // not inline us.
+                    if self.source.def_id() == callsite.callee {
+                        // Bail on direct recursion
+                        continue 'callsites;
                     }
-                } else {
-                    // This cannot result in a cycle since the callee MIR is from another crate
-                    // and is already optimized.
-                    self.tcx.optimized_mir(callsite.callee)
-                };
+                    let mut seen: FxHashSet<_> = Default::default();
+                    let mut todo = vec![callsite.callee];
+                    while let Some(check) = todo.pop() {
+                        seen.insert(check);
+                        let local_calls = self.tcx.local_calls(check);
+                        for &d in local_calls {
+                            if d == self.source.def_id() {
+                                continue 'callsites;
+                            }
+                            if seen.insert(d) {
+                                todo.push(d);
```

Comment:
Not sure if it matters, but `todo` could also be a `VecDeque` instead of a `Vec`.

Reply:
That just changes the order in which these are processed. If we are talking order, we should figure out some heuristic for which DefIds are most likely going to cause a cycle. We can guess this from DefIds because of the order that we process things in. Although in a pure query based system this wouldn't work anymore.